src/cpg_flow/metamist.py

```diff
@@ -0,0 +1,617 @@
+        """
+        try:
+            return api_func(**kwargv)
+        except ServiceException:
+            # raise here so the retry occurs
+            logging.warning(
+                f"Retrying {api_func} ...",
+            )
+            raise
+
+    def make_aapi_call(self, api_func: Callable, **kwargv: Any):
+        """
+        Make a generic API call to self.aapi.
+        This is a wrapper around retry of API call to handle exceptions and logging.
+        """
+        try:
+            return self.make_retry_aapi_call(api_func, **kwargv)
+        except (ServiceException, ApiException) as e:
+            # Metamist API failed even after retries
+            # log the error and continue
+            traceback.print_exc()
+            logging.error(
+                f"Error: {e} Call {api_func} failed with payload:\n{str(kwargv)}",
+            )
+        # TODO: discuss should we catch all here as well?
+        # except Exception as e:
+        #     # Other exceptions?
+
+        return None
+
+    @retry(
+        stop=stop_after_attempt(3),
+        wait=wait_exponential(multiplier=3, min=8, max=30),
+        retry=retry_if_exception_type(TransportServerError),
+        reraise=True,
+    )
+    def get_sg_entries(self, dataset_name: str) -> list[dict]:
+        """
+        Retrieve sequencing group entries for a dataset, in the context of access level
+        and filtering options.
+        """
+        metamist_proj = self.get_metamist_proj(dataset_name)
+        logging.info(f"Getting sequencing groups for dataset {metamist_proj}")
+
+        skip_sgs = get_config()["workflow"].get("skip_sgs", [])
+        only_sgs = get_config()["workflow"].get("only_sgs", [])
+        sequencing_type = get_config()["workflow"].get("sequencing_type")
+
+        if only_sgs and skip_sgs:
+            raise MetamistError("Cannot specify both only_sgs and skip_sgs in config")
+
+        sequencing_group_entries = query(
+            GET_SEQUENCING_GROUPS_QUERY,
+            variables={
+                "metamist_proj": metamist_proj,
+                "only_sgs": only_sgs,
+                "skip_sgs": skip_sgs,
+                "sequencing_type": sequencing_type,
+            },
+        )
+
+        return sequencing_group_entries["project"]["sequencingGroups"]
+
+    def update_analysis(self, analysis: Analysis, status: AnalysisStatus):
+        """
+        Update "status" of an Analysis entry.
+        """
+        self.make_aapi_call(
+            self.aapi.update_analysis,
+            analysis_id=analysis.id,
+            analysis_update_model=models.AnalysisUpdateModel(
+                status=models.AnalysisStatus(status.value),
+            ),
+        )
+        # Keeping this as is for compatibility with the existing code
+        # However this should only be set after the API call is successful
+        analysis.status = status
+
+    # TODO: Refactor based on correct usage
```

Comment:
Did we handle this todo already? It looks like we removed those un-used params. Is that ok given where it is used?

Reply:
SY/YP: We'll leave this TODO for the MVP phase for now